fs-storage/src/file_storage.rs

```diff
@@ -105,26 +105,36 @@ where
         self.data.entries.remove(id).ok_or_else(|| {
             ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
         })?;
-        self.timestamp = std::time::SystemTime::now();
+        self.modified = std::time::SystemTime::now();
         self.write_fs()
             .expect("Failed to remove data from disk");
         Ok(())
     }
 
     /// Compare the timestamp of the storage file with the timestamp of the storage instance
-    /// to determine if the storage file has been updated.
-    fn is_storage_updated(&self) -> Result<bool> {
-        let file_timestamp = fs::metadata(&self.path)?.modified()?;
-        let file_time_secs = file_timestamp
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .unwrap()
-            .as_secs();
-        let self_time_secs = self
-            .timestamp
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .unwrap()
-            .as_secs();
-        Ok(file_time_secs > self_time_secs)
+    /// to determine if the storage file is up-to-date.
+    fn is_outdated(&self) -> Result<bool> {
+        match fs::metadata(&self.path) {
+            Ok(metadata) => {
+                let fs_modified = metadata
+                    .modified()?
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs();
+
+                let self_modified = self
+                    .modified
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs();
+
+                Ok(fs_modified != self_modified)
+            },
+            Err(e) => {
+                //todo: log the error
```

Comment:
@Pushkarm029 could you double-check my refactoring of this function, optimize if possible and add error logs? This function was not really correct before, and to make it completely correct we need [atomic versions](https://github.com/ARK-Builders/ark-rust/issues/3) but at least we should flag correctly when the data doesn't match the disk. This can be either because of updates to the in-memory map, or because of updates to the filesystem.

Reply:
Using metadata to check outdatedness is a valid solution. In fact, mature tools like rsync do the same. Perhaps we can increase the time resolution to use milliseconds with `as_millis`.

Also about atomic versions, I think it brings in significant complexity to the implementation so it can be put behind a feature flag. More, many use cases that involve user interaction it is very unlikely for their to be race conditions related to timestamps. Such use cases can use this (default) implementation. For automated updated use cases, the feature flag can be turned on and use atomic versions.

Reply:
Yes, atomic versions is more complicated mechanism. We don't need to implement it in this PR, but it will be a good thing.

Rsync is a bit different use-case if look at it on single-machine scale, because it only reads state of the filesystem (write is happening on a remote machine). And we do both reading and writing within single machine. But I think timestamp solution is good enough for now, the only problem is that we only can detect when we need an update, but we can't figure out which side (fs or ram) is behind the other. Because changes could be made on both sides.

But atomic versions probably don't solve this case, we would still have diverged versions and sync them using monoids 🤔